src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs

```diff
@@ -395,10 +395,22 @@ public Task PublishUpdateAsync(IResource resource, string resourceId, Func<Custo
     /// </summary>
     private static CustomResourceSnapshot UpdateHealthStatus(IResource resource, CustomResourceSnapshot previousState)
     {
-        // A resource is also healthy if it has no health check annotations and is in the running state.
-        if (previousState.HealthStatus is not HealthStatus.Healthy && !resource.TryGetAnnotationsIncludingAncestorsOfType<HealthCheckAnnotation>(out _) && previousState.State?.Text == KnownResourceStates.Running)
+        // A resource is also healthy if it
+        // - has a null health status (wasn't set by a health check nor initial snapshot)
+        // - has no health check annotations
+        // - is in the running state
+        // - was not started with an initial snapshot
+        if (previousState.HealthStatus is null)
         {
-            return previousState with { HealthStatus = HealthStatus.Healthy };
+            var hasHealthChecks = resource.TryGetAnnotationsIncludingAncestorsOfType<HealthCheckAnnotation>(out _);
+            var hasInitialSnapshot = resource.TryGetLastAnnotation<ResourceSnapshotAnnotation>(out _);
+
+            if (!hasHealthChecks
+                && previousState.State?.Text == KnownResourceStates.Running
```

Comment:
nit: Check at the same time as health status. It is cheaper that getting annotations.